buffer/src/main/java/io/netty/buffer/ByteBufUtil.java

```diff
@@ -352,7 +616,527 @@ static String decodeString(ByteBuffer src, Charset charset) {
+     * Returns a multi-line hexadecimal dump of the specified {@link ByteBuf} that is easy to read by humans.
+     */
+    public static String prettyHexDump(ByteBuf buffer) {
+        return prettyHexDump(buffer, buffer.readerIndex(), buffer.readableBytes());
+    }
+
+    /**
+     * Returns a multi-line hexadecimal dump of the specified {@link ByteBuf} that is easy to read by humans,
+     * starting at the given {@code offset} using the given {@code length}.
+     */
+    public static String prettyHexDump(ByteBuf buffer, int offset, int length) {
+        return HexUtil.prettyHexDump(buffer, offset, length);
+    }
+
+    /**
+     * Appends the prettified multi-line hexadecimal dump of the specified {@link ByteBuf} to the specified
+     * {@link StringBuilder} that is easy to read by humans.
+     */
+    public static void appendPrettyHexDump(StringBuilder dump, ByteBuf buf) {
+        appendPrettyHexDump(dump, buf, buf.readerIndex(), buf.readableBytes());
+    }
+
+    /**
+     * Appends the prettified multi-line hexadecimal dump of the specified {@link ByteBuf} to the specified
+     * {@link StringBuilder} that is easy to read by humans, starting at the given {@code offset} using
+     * the given {@code length}.
+     */
+    public static void appendPrettyHexDump(StringBuilder dump, ByteBuf buf, int offset, int length) {
+        HexUtil.appendPrettyHexDump(dump, buf, offset, length);
+    }
+
+    /* Separate class so that the expensive static initialization is only done when needed */
+    private static final class HexUtil {
+
+        private static final char[] BYTE2CHAR = new char[256];
+        private static final char[] HEXDUMP_TABLE = new char[256 * 4];
+        private static final String[] HEXPADDING = new String[16];
+        private static final String[] HEXDUMP_ROWPREFIXES = new String[65536 >>> 4];
+        private static final String[] BYTE2HEX = new String[256];
+        private static final String[] BYTEPADDING = new String[16];
+
+        static {
+            final char[] DIGITS = "0123456789abcdef".toCharArray();
+            for (int i = 0; i < 256; i ++) {
+                HEXDUMP_TABLE[ i << 1     ] = DIGITS[i >>> 4 & 0x0F];
+                HEXDUMP_TABLE[(i << 1) + 1] = DIGITS[i       & 0x0F];
+            }
+
+            int i;
+
+            // Generate the lookup table for hex dump paddings
+            for (i = 0; i < HEXPADDING.length; i ++) {
+                int padding = HEXPADDING.length - i;
+                StringBuilder buf = new StringBuilder(padding * 3);
+                for (int j = 0; j < padding; j ++) {
+                    buf.append("   ");
+                }
+                HEXPADDING[i] = buf.toString();
+            }
+
+            // Generate the lookup table for the start-offset header in each row (up to 64KiB).
+            for (i = 0; i < HEXDUMP_ROWPREFIXES.length; i ++) {
+                StringBuilder buf = new StringBuilder(12);
+                buf.append(NEWLINE);
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) StringBuffer (or StringBuilder).append is called consecutively without reusing the target variable. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=pmd%3AConsecutiveAppendsShouldReuse)